src/wp-includes/class-wp-scripts.php

```diff
@@ -714,6 +789,234 @@ public function in_default_dir( $src ) {
+	 */
+	private function is_valid_strategy( $strategy ) {
+		$allowed_strategies = array( 'blocking', 'defer', 'async' );
+
+		return in_array( $strategy, $allowed_strategies, true );
+	}
+
+	/**
+	 * Get the strategy assigned during script registration.
+	 *
+	 * @param string $handle The script handle.
+	 * @return string|bool Strategy set during script registration. False if none was set.
+	 */
+	private function get_intended_strategy( $handle ) {
+		$script_args = $this->get_data( $handle, 'script_args' );
+		$strategy    = isset( $script_args['strategy'] ) ? $script_args['strategy'] : false;
+
+		if ( $strategy && ! $this->is_valid_strategy( $strategy ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				sprintf(
+					/* translators: 1: $strategy, 2: $handle */
+					__( 'Invalid strategy `%1$s` defined for `%2$s` during script registration.' ),
+					$strategy,
+					$handle
+				),
+				'6.3.0'
+			);
+
+			return false;
+		}
+
+		return $strategy;
+	}
+
+	/**
+	 * Check if the strategy passed is a non-blocking strategy.
+	 *
+	 * @param string $strategy The strategy to check.
+	 * @return bool True if $strategy is one of the delayed strategy.
+	 */
+	private function is_non_blocking_strategy( $strategy ) {
+		$delayed_strategies = array( 'defer', 'async' );
+
+		return in_array( $strategy, $delayed_strategies, true );
+	}
+
+	/**
+	 * Check if a script has a non standalone inline script associated with it.
+	 *
+	 * @param string $handle   The script handle.
+	 * @param string $position Position of the inline script.
+	 *
+	 * @return bool True if script present. False if empty.
+	 */
+	private function has_non_standalone_inline_script( $handle, $position ) {
+		$non_standalone_script_key = $position . '-non-standalone';
+		$non_standalone_script     = $this->get_data( $handle, $non_standalone_script_key );
+
+		return ! empty( $non_standalone_script );
+	}
+
+	/**
+	 * Check if all of a scripts dependents are deferrable, which is required to maintain execution order.
+	 *
+	 * @param string $handle  The script handle.
+	 * @param array  $checked Optional. An array of already checked script handles, used to avoid recursive loops.
+	 * @return bool True if all dependents are deferrable, false otherwise.
+	 */
+	private function has_only_deferrable_dependents( $handle, $checked = array() ) {
+		// If this node was already checked, this script can be deferred and the branch ends.
+		if ( in_array( $handle, $checked, true ) ) {
+			return true;
+		}
+
+		$checked[]  = $handle;
+		$dependents = $this->get_dependents( $handle );
+
+		// If there are no dependents remaining to consider, the script can be deferred.
+		if ( empty( $dependents ) ) {
+			return true;
+		}
+
+		// Consider each dependent and check if it is deferrable.
+		foreach ( $dependents as $dependent ) {
+			// If the dependent script is not using the defer or async strategy, no script in the chain is deferrable.
+			$strategy = $this->get_intended_strategy( $dependent );
```

Comment:
in some cases this could return `false`, in which case we could exit early, right?